### ISSUE TITLE
Normal AI aggressive only player clan ❓ 

### ISSUE THREAD

**Issue by IlnurNasybullin** [SMUnlimited/AMAI#360]
I played on map "Divide and Conquer" in mode 3x3x3x3 (one player and 11 Normal AI). And every time within 2-3 minutes, 2 clans (3 bots each) attacked my clan from two sides and in fact, the game came down to when my clan was completely finished off. I checked the map (turned off the fog of war) and the last clan did not make any attacks on neighboring clans - it calmly crept. Is that how it's meant to be? Don't bots have too aggressive style (and don't look very human)? It seemed to me that bots try to behave as much as possible as if the player were playing against another person

**Comment by jzy-chitong56** [comment]
In fact, there is such an attack mode in my branch. Essentially, AI defeating human players is a victory, and internal struggles are not conducive to this.  should unite against the human

But I think they might start fighting at your doorstep

**Comment by SMUnlimited** [comment]
Its partly using the existing hardcoded system to decide major attacks at the moment. Otherwise random element to attack the nearest and weakest enemy.

And if one targets you its allies can also help attack the same target.

**Comment by paulo101977** [comment]
> Its partly using the existing hardcoded system to decide major attacks at the moment. Otherwise random element to attack the nearest and weakest enemy.
> 
> And if one targets you its allies can also help attack the same target.

The only way to fix it is as I had thought before: Reinforcement Learning. However, as the game does not allow access to data from the running game, it is difficult to create an efficient mechanism for Machine Learning. I even thought about using the game's memory, but I would have to have everything mapped out and a way to communicate the information back.

**Comment by jzy-chitong56** [comment]
> > Its partly using the existing hardcoded system to decide major attacks at the moment. Otherwise random element to attack the nearest and weakest enemy.
> > And if one targets you its allies can also help attack the same target.
> 
> The only way to fix it is as I had thought before: Reinforcement Learning. However, as the game does not allow access to data from the running game, it is difficult to create an efficient mechanism for Machine Learning. I even thought about using the game's memory, but I would have to have everything mapped out and a way to communicate the information back.

This is one of the main reasons why China is stuck in the old version, and there is no one
In versions 1.27a or below, they exploited RB vulnerabilities and some underlying implants to create third-party APIs
These APIs were originally designed to facilitate authors, such as implementing complex functions or skill effects with a single trigger action, easily excreting variables, and accessing low-level APIs that Blizzard has not yet opened

But later on, the battle platform integrated some APIs and implemented a membership system and an in app purchasing system. For example, by purchasing a membership, users can obtain advanced props, skills, or skins for certain maps, which are effective every time or have the right to run certain maps. They can even interact in real-time with live broadcasts and audience comments
In other words, in China, lower versions can realize your ideas
But what is the cost

**Comment by IlnurNasybullin** [comment]
> Частично это использование существующей жестко закодированной системы для определения основных атак на данный момент. В противном случае случайный элемент для атаки ближайшего и самого слабого врага.

As I understand it, this cannot be corrected in any way (at least in a simple way)? And another question - is how well, in general, this AMAI bot is suitable for FFA team games (when there are more than 2 clans in game)?

**Comment by IlnurNasybullin** [comment]
> In fact, there is such an attack mode in my branch. Essentially, AI defeating human players is a victory, and internal struggles are not conducive to this. should unite against the human

As I understand it, this cannot be corrected in any way (at least in a simple way)? And another question - is how well, in general, this AMAI bot is suitable for FFA team games (when there are more than 2 clans in game)?

**Comment by SMUnlimited** [comment]
> > In fact, there is such an attack mode in my branch. Essentially, AI defeating human players is a victory, and internal struggles are not conducive to this. should unite against the human
> 
> As I understand it, this cannot be corrected in any way (at least in a simple way)? And another question - is how well, in general, this AMAI bot is suitable for FFA team games (when there are more than 2 clans in game)?

We can make them attack whoever we want, we would have to drop hardcoded logic though, then its just working out the real right target at the right moment which is complex. Map type, distance, strength, allies, current army locations, current strategy, enemy strategy, target location, journey over map, are just some factors we have to consider in a new attack function.

**Comment by IlnurNasybullin** [comment]
Ok, thanks, maybe if I'm really interested in this topic, I'll explore the script code.
I'm close this issue

**Comment by IlnurNasybullin** [comment]
Ok, thanks, maybe if I'm really interested in this topic, I'll explore the script code.
I'm close this issue

**Comment by paulo101977** [comment]
> > > Its partly using the existing hardcoded system to decide major attacks at the moment. Otherwise random element to attack the nearest and weakest enemy.
> > > And if one targets you its allies can also help attack the same target.
> > 
> > 
> > The only way to fix it is as I had thought before: Reinforcement Learning. However, as the game does not allow access to data from the running game, it is difficult to create an efficient mechanism for Machine Learning. I even thought about using the game's memory, but I would have to have everything mapped out and a way to communicate the information back.
> 
> This is one of the main reasons why China is stuck in the old version, and there is no one In versions 1.27a or below, they exploited RB vulnerabilities and some underlying implants to create third-party APIs These APIs were originally designed to facilitate authors, such as implementing complex functions or skill effects with a single trigger action, easily excreting variables, and accessing low-level APIs that Blizzard has not yet opened
> 
> But later on, the battle platform integrated some APIs and implemented a membership system and an in app purchasing system. For example, by purchasing a membership, users can obtain advanced props, skills, or skins for certain maps, which are effective every time or only platform have the right to run certain maps. They can even interact in real-time with live broadcasts and audience comments Of course, all income must be shared by the platform In other words, in China, lower versions can realize your ideas But what is the cost

That's the problem, being stuck with an old version. If Blizzard provided an API that would allow me to access the game in real time, everything would make things easier. The problem is that this would also make it easier for cheaters, etc. And there is still the issue of cost. But in the case of a two-way API, in which I receive and send information to/from the game, it would be useful and the cost of producing an RL algorithm would be much lower.
